### ISSUE TITLE
ZFP compression for Blueprint HDF5 extracts?

### ISSUE THREAD

**Issue by BenWibking** [Alpine-DAV/ascent#1171]
Is there a way to turn on ZFP compression for Blueprint HDF5 extracts? Or gzip compression?

**Comment by cyrush** [comment]
GZip + HDF5 should be supported out of the box, as long as HDF5 was built with zlib support. (build_ascent.sh does build hdf5 with zlib)

GZip use is triggered by a set of size heuristics, on a per-array basis during HDF5 writes. 

The knobs that control those size heuristics aren't wired up to the blueprint extract, but we could add them -- we have a ticket for this:

https://github.com/Alpine-DAV/ascent/issues/902

Also, here is an example some of those thresholds as used in conduit directly to give you an idea of what controls exist:
https://llnl-conduit.readthedocs.io/en/latest/relay_io.html#hdf5-i-o-options


For ZFP, we have future plans to support for Blueprint Meshes, but it's not supported yet. 

ZFP works best when it understands the ND-array nature of the data, there is some work to plumb that for general blueprint meshes.

**Comment by BenWibking** [comment]
Adding parameters for the size heuristics would be useful. Is the chunking threshold in bytes, or counts?

My extracts aren't getting chunked or compressed, but there is a significant compression ratio when I manually gzip them.

```
Dataset layout information:
	Dataset layout counts[COMPACT]: 682
	Dataset layout counts[CONTIG]: 352
	Dataset layout counts[CHUNKED]: 0
	Dataset layout counts[VIRTUAL]: 0
	Number of external files : 0
Dataset filters information:
	Number of datasets with:
		NO filter: 1034
		GZIP filter: 0
		SHUFFLE filter: 0
		FLETCHER32 filter: 0
		SZIP filter: 0
		NBIT filter: 0
		SCALEOFFSET filter: 0
		USER-DEFINED filter: 0
```

**Comment by cyrush** [comment]
Good question -- the chunking threshold (and similarly the compact threshold) is specified as a number of bytes (not elements).

One complexity is that HDF5 chunking itself, which is required to use the compression filters, introduces overhead in the HDF5 book keeping data structures. So there is a trade off, and gziping the file could still yield gains as it will also work on the HDF5 metadata. 

I'll try to get these wired up this week so you can experiment.

**Comment by cyrush** [comment]
@BenWibking  I merged a PR that provides these options in #1173

Added some details to the relay extract docs:

https://ascent.readthedocs.io/en/latest/Actions/Extracts.html#relay

And here is a unit test that exercises them:

https://github.com/Alpine-DAV/ascent/blob/b8084a13b8a0569783b81543ff60f5425291383a/src/tests/ascent/t_ascent_relay.cpp#L163

Will close this issue & I'll keep you in the loop if there is progress on ZFP support.
